pyro/infer/abstract_infer.py

```diff
@@ -326,7 +326,10 @@ def _adjust_to_data(self, trace, data_trace):
                     # Select random sub-indices to replay values under conditionally independent stacks.
                     # Otherwise, we assume there is an dependence of indexes between training data
                     # and prediction data.
-                    batch_dim = cis.dim - site["fn"].event_dim
+                    if site_is_subsample(site):
+                        batch_dim = cis.dim
```

Comment:
I don't think we should need this because subsample sites shouldn't have non-empty `cond_indep_stack`. I'm also noticing that we can remove the try-catch block by only iterating over `sample` sites to begin with, something like this:

```diff
     def _adjust_to_data(self, trace, data_trace):
-        for name, site in list(trace.nodes.items()):
+        for name, site in trace.iter_stochastic_nodes():
             # Adjust subsample sites
             if site_is_subsample(site):
                 site["fn"] = data_trace.nodes[name]["fn"]
                 site["value"] = data_trace.nodes[name]["value"]
             # Adjust sites under conditionally independent stacks
-            try:
-                site["cond_indep_stack"] = data_trace.nodes[name]["cond_indep_stack"]
-                site["fn"] = data_trace.nodes[name]["fn"]
-                for cis in site["cond_indep_stack"]:
-                    # Select random sub-indices to replay values under conditionally independent stacks.
-                    # Otherwise, we assume there is an dependence of indexes between training data
-                    # and prediction data.
-                    if site_is_subsample(site):
-                        batch_dim = cis.dim
-                    else:
-                        batch_dim = cis.dim - site["fn"].event_dim
-                    subidxs = torch.randint(0, site['value'].size(batch_dim), (cis.size,),
-                                            device=site["value"].device)
-                    site["value"] = site["value"].index_select(batch_dim, subidxs)
-            except KeyError:
-                pass
+            site["cond_indep_stack"] = data_trace.nodes[name]["cond_indep_stack"]
+            site["fn"] = data_trace.nodes[name]["fn"]
+            for cis in site["cond_indep_stack"]:
+                # Select random sub-indices to replay values under conditionally independent stacks.
+                # Otherwise, we assume there is an dependence of indexes between training data
+                # and prediction data.
+                batch_dim = cis.dim - site["fn"].event_dim
+                subidxs = torch.randint(0, site['value'].size(batch_dim), (cis.size,),
+                                        device=site["value"].device)
+                site["value"] = site["value"].index_select(batch_dim, subidxs)
```

Do you think this will work for all your use cases?

Reply:
Yes, that will cover my use cases! 😄